### PR TITLE
Dedicated method for creating circuit from op tree with EARLIEST strategy

### PR DESCRIPTION
**Pull request by daxfohl**
We keep track of the latest moment that contains each qubit and measurement key. Then we know where to put each operation in constant time. We don't create the actual Moments until the very end, when we know where everything goes. Also added explicit key protocol impls for EigenGate, preempting the protocol from attempting a bunch of fallback options. On my laptop this speeds up creating circuits with EARLIEST strategy by almost infinity percent. (On my laptop, the circuit in the new test goes from 29.00s on master to 0.13s here).

### CODE REVIEW

File: cirq-core/cirq/circuits/circuit.py [hunk 1]
```diff
@@ -1704,7 +1704,47 @@ def __init__(
         """
         self._moments: List['cirq.Moment'] = []
         with _compat.block_overlapping_deprecation('.*'):
-            self.append(contents, strategy=strategy)
+            if strategy == InsertStrategy.EARLIEST:
+                self._create_from_earliest(contents)
+            else:
+                self.append(contents, strategy=strategy)
+
+    def _create_from_earliest(self, contents):
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/circuits/circuit.py]: Added comments and types. LMK what you think.

**Review by dabacon — COMMENTED** [review]
Very nice!

**Comment by daxfohl** [comment]
@dstrain115 I updated the documentation. Should this be merged before 0.15?

**Review by dstrain115 — APPROVED** [review]
Sorry, I lost track of this.  Let's merge this!

**Comment by daxfohl** [comment]
Merge label? I don't have authz

File: cirq-core/cirq/circuits/circuit.py [hunk 2]
```diff
@@ -1973,14 +2061,9 @@ def insert(
...
             ValueError: Bad insertion strategy.
         """
-        moments_and_operations = list(
-            ops.flatten_to_ops_or_moments(
-                ops.transform_op_tree(moment_or_operation_tree, preserve_moments=True)
-            )
-        )
         # limit index to 0..len(self._moments), also deal with indices smaller 0
         k = max(min(index if index >= 0 else len(self._moments) + index, len(self._moments)), 0)
-        for moment_or_op in moments_and_operations:
```
Comment (vtomole) [inline cirq-core/cirq/circuits/circuit.py:1983]: Link

**Pull request merged**